community/kernel/src/main/java/org/neo4j/kernel/impl/store/TemporalValueWriterAdapter.java

```diff
@@ -0,0 +1,152 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.store;
+
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+import java.time.OffsetTime;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+
+import org.neo4j.values.storable.ValueWriter;
+import org.neo4j.values.utils.TemporalUtil;
+
+import static java.time.ZoneOffset.UTC;
+
+/**
+ * A {@link ValueWriter} that defines format for all temporal types, except duration.
+ * Subclasses will not be able to override methods like {@link #writeDate(LocalDate)}. They should instead override {@link #writeDate(long)} that
+ * defines how {@link LocalDate} is serialized.
+ * <p>
+ * Primary purpose of this class is to share serialization format between property store writer and schema indexes.
+ *
+ * @param <E> the error type.
+ */
+public abstract class TemporalValueWriterAdapter<E extends Exception> extends ValueWriter.Adapter<E>
+{
+    @Override
+    public final void writeDate( LocalDate localDate ) throws E
+    {
+        writeDate( localDate.toEpochDay() );
+    }
+
+    @Override
+    public final void writeLocalTime( LocalTime localTime ) throws E
+    {
+        writeLocalTime( localTime.toNanoOfDay() );
+    }
+
+    @Override
+    public final void writeTime( OffsetTime offsetTime ) throws E
+    {
+        long nanosOfDayUTC = TemporalUtil.getNanosOfDayUTC( offsetTime );
+        int offsetSeconds = offsetTime.getOffset().getTotalSeconds();
+        writeTime( nanosOfDayUTC, offsetSeconds );
+    }
+
+    @Override
+    public final void writeLocalDateTime( LocalDateTime localDateTime ) throws E
+    {
+        long epochSecond = localDateTime.toEpochSecond( UTC );
+        int nano = localDateTime.getNano();
+        writeLocalDateTime( epochSecond, nano );
+    }
+
+    @Override
+    public final void writeDateTime( ZonedDateTime zonedDateTime ) throws E
+    {
+        long epochSecondUTC = zonedDateTime.toEpochSecond();
+        int nano = zonedDateTime.getNano();
+
+        ZoneId zone = zonedDateTime.getZone();
+        if ( zone instanceof ZoneOffset )
+        {
+            int offsetSeconds = ((ZoneOffset) zone).getTotalSeconds();
+            writeDateTime( epochSecondUTC, nano, offsetSeconds );
+        }
+        else
+        {
+            String zoneId = zone.getId();
+            writeDateTime( epochSecondUTC, nano, zoneId );
+        }
+    }
+
+    /**
+     * Write date value obtained from {@link LocalDate} in {@link #writeDate(LocalDate)}.
+     *
+     * @param epochDay the epoch day.
+     */
+    protected void writeDate( long epochDay ) throws E
```

Comment:
What is the reason this method and the ones below are not abstract?

Reply:
They live in an "adaptor" class and it's subclasses (those who extend `NativeSchemaKey`) usually decide to override just one method. So methods are not abstract to not force multiple empty/throwing overrides in subclasses.